### DISCUSSION TITLE
Best practices for multiple large header files?

### DISCUSSION THREAD

**Discussion by Supreeeme** [rust-lang/rust-bindgen#2915] [Q&A]
I have several (5-10+) ~6000 line headers I want to generate bindings for. (These headers have a lot of the same types but I post-process the bindings from bindgen to deduplicate them.) My question is what is the best practice to efficiently generate bindings for these files? Obviously I can stick them all in one bindgen call, but that seems to really slow down as I add headers. My second thought was to split up the bindgen invocations and run them in parallel, but with fewer files this seems slower than just doing it all in one call.

Is there some kind of best number of lines/files where it makes sense to split invocations across threads? Or is bindgen already multi threaded and this is pointless?

**Comment by pvdrz** [comment]
I haven't read any discussion regarding the performance of bindgen other than the impact of enabling the attribute detection feature which slowed bindgen considerably on larger C++ headers.

If you're deduplicating a lot of items you could use the `allowlist_.*`/`blocklist_.*` methods to avoid generating bindings for those on the first place and that could save you some time.

Maybe it would make sense to profile bindgen and check where are we spending more time and why.

**Reply by Supreeeme** [reply]
What would be the best way to profile it from a build script?

**Reply by pvdrz** [reply]
There's https://docs.rs/bindgen/latest/bindgen/struct.Builder.html#method.time_phases which prints some timings to stderr already. If you need anything more granular, you could use `perf` (see: https://nnethercote.github.io/perf-book/profiling.html)
